File: BASS_LIBS/bass24-osx 2/rectest/rectest.c
/*
	BASS recording example
	Copyright (c) 2002-2019 Un4seen Developments Ltd.
*/

#include <Carbon/Carbon.h>
#include "bass.h"

WindowPtr win;

#define FREQ 44100
#define CHANS 2
#define BUFSTEP 200000	// memory allocation unit

int input;				// current input source
char *recbuf = NULL;		// recording buffer
DWORD reclen;			// recording length

HRECORD rchan = 0;		// recording channel
HSTREAM chan = 0;			// playback channel

// display error messages
void Error(const char *es)
{
	short i;
	char mes[200];
	sprintf(mes, "%s\n(error code: %d)", es, BASS_ErrorGetCode());
	CFStringRef ces = CFStringCreateWithCString(0, mes, 0);
	DialogRef alert;
	CreateStandardAlert(0, CFSTR("Error"), ces, NULL, &alert);
	RunStandardAlert(alert, NULL, &i);
	CFRelease(ces);
}

ControlRef GetControl(int id)
{
	ControlRef cref;
	ControlID cid = { 0,id };
	GetControlByID(win, &cid, &cref);
	return cref;
}

void SetupControlHandler(int id, DWORD event, EventHandlerProcPtr proc)
{
	EventTypeSpec etype = { kEventClassControl,event };
	ControlRef cref = GetControl(id);
	InstallControlEventHandler(cref, NewEventHandlerUPP(proc), 1, &etype, cref, NULL);
}

void SetControlText(int id, const char *text)
{
	CFStringRef cs = CFStringCreateWithCString(0, text, 0);
	SetControlTitleWithCFString(GetControl(id), cs);
	CFRelease(cs);
}

// buffer the recorded data
BOOL CALLBACK RecordingCallback(HRECORD handle, const void *buffer, DWORD length, void *user)
{
	// increase buffer size if needed
	if ((reclen % BUFSTEP) + length >= BUFSTEP) {
		recbuf = realloc(recbuf, ((reclen + length) / BUFSTEP + 1) * BUFSTEP);
		if (!recbuf) {
			rchan = 0;
//			Error("Out of memory!");
			SetControlText(10, "Record");
			return FALSE; // stop recording
		}
	}
	// buffer the data
	memcpy(recbuf + reclen, buffer, length);
	reclen += length;
	return TRUE; // continue recording
}

void StartRecording()
{
	WAVEFORMATEX *wf;
	if (recbuf) { // free old recording
		BASS_StreamFree(chan);
		chan = 0;
		free(recbuf);
		recbuf = NULL;
		DisableControl(GetControl(11));
		DisableControl(GetControl(12));
	}
	// allocate initial buffer and make space for WAVE header
	recbuf = malloc(BUFSTEP);
	reclen = 44;
	// fill the WAVE header
	memcpy(recbuf, "RIFF\0\0\0\0WAVEfmt \20\0\0\0", 20);
	memcpy(recbuf + 36, "data\0\0\0\0", 8);
	wf = (WAVEFORMATEX*)(recbuf + 20);
	wf->wFormatTag = 1;
	wf->nChannels = CHANS;
	wf->wBitsPerSample = 16;
	wf->nSamplesPerSec = FREQ;
	wf->nBlockAlign = wf->nChannels * wf->wBitsPerSample / 8;
	wf->nAvgBytesPerSec = wf->nSamplesPerSec * wf->nBlockAlign;
	// start recording
	rchan = BASS_RecordStart(FREQ, CHANS, 0, RecordingCallback, 0);
	if (!rchan) {
		Error("Can't start recording");
		free(recbuf);
		recbuf = 0;
		return;
	}
	SetControlText(10, "Stop");
}

void StopRecording()
{
	BASS_ChannelStop(rchan);
	rchan = 0;
	SetControlText(10, "Record");
	// complete the WAVE header
	*(DWORD*)(recbuf + 4) = reclen - 8;
	*(DWORD*)(recbuf + 40) = reclen - 44;
	// enable "save" button
	EnableControl(GetControl(12));
	// create a stream from the recording
	if (chan = BASS_StreamCreateFile(TRUE, recbuf, 0, reclen, 0))
		EnableControl(GetControl(11)); // enable "play" button
}

// write the recorded data to disk
void WriteToDisk()
{
	NavDialogRef fileDialog;
	NavDialogCreationOptions fo;
	NavGetDefaultDialogCreationOptions(&fo);
	fo.optionFlags = 0;
	fo.parentWindow = win;
	fo.saveFileName = CFSTR("bass.wav");
	NavCreatePutFileDialog(&fo, 0, 0, NULL, NULL, &fileDialog);
	if (!NavDialogRun(fileDialog)) {
		NavReplyRecord r;
		if (!NavDialogGetReply(fileDialog, &r)) {
			AEKeyword k;
			FSRef fr;
			if (!AEGetNthPtr(&r.selection, 1, typeFSRef, &k, NULL, &fr, sizeof(fr), NULL)) {
				FILE *fp;
				char file[256], *fe;
				FSRefMakePath(&fr, (BYTE*)file, sizeof(file));
				fe = strchr(file, 0);
				*fe++ = '/';
				CFStringGetCString(r.saveFileName, fe, file + sizeof(file) - fe, kCFStringEncodingUTF8);
				if (fp = fopen(file, "wb")) {
					fwrite(recbuf, reclen, 1, fp);
					fclose(fp);
				} else
					Error("Can't create the file");
			}
			NavDisposeReply(&r);
		}
	}
	NavDialogDispose(fileDialog);
}

void UpdateInputInfo()
{
	float level;
	int it = BASS_RecordGetInput(input, &level); // get info on the input
	if (it == -1 || level < 0) { // failed to get level
		level = 1; // just display 100%
		DisableControl(GetControl(14));
	} else
		EnableControl(GetControl(14));
	SetControl32BitValue(GetControl(14), level * 100); // set the level slider
}

pascal OSStatus RecordEventHandler(EventHandlerCallRef inHandlerRef, EventRef inEvent, void *inUserData)
{
	if (!rchan)
		StartRecording();
	else
		StopRecording();
	return noErr;
}

pascal OSStatus PlayEventHandler(EventHandlerCallRef inHandlerRef, EventRef inEvent, void *inUserData)
{
	BASS_ChannelPlay(chan, TRUE); // play the recorded data
	return noErr;
}

pascal OSStatus SaveEventHandler(EventHandlerCallRef inHandlerRef, EventRef inEvent, void *inUserData)
{
	WriteToDisk();
	return noErr;
}

pascal OSStatus DeviceEventHandler(EventHandlerCallRef inHandlerRef, EventRef inEvent, void *inUserData)
{
	input = GetControl32BitValue(inUserData) - 1; // get the selection
	BASS_RecordSetInput(input, BASS_INPUT_ON, -1); // enable the selected input
	UpdateInputInfo(); // update info
	return noErr;
}

pascal void VolumeEventHandler(ControlHandle control, SInt16 part)
{ // set input source level
	float level = GetControl32BitValue(control) / 100.f;
	BASS_RecordSetInput(input, 0, level);
}

pascal void TimerProc(EventLoopTimerRef inTimer, void *inUserData)
{ // update the recording/playback counter
	char text[30] = "";
	if (rchan) { // recording
		if (!BASS_ChannelIsActive(rchan)) { // the recording has stopped, eg. unplugged device
			StopRecording();
			Error("The recording stopped");
			return;
		}
		sprintf(text, "%d", reclen - 44);
	} else if (chan) {
		if (BASS_ChannelIsActive(chan)) // playing
			sprintf(text, "%lld / %lld", BASS_ChannelGetPosition(chan, BASS_POS_BYTE), BASS_ChannelGetLength(chan, BASS_POS_BYTE));
		else
			sprintf(text, "%lld", BASS_ChannelGetLength(chan, BASS_POS_BYTE));
	}
	{
		ControlRef cref = GetControl(20);
		SetControlData(cref, kControlNoPart, kControlStaticTextTextTag, strlen(text), text);
		DrawOneControl(cref);
	}
}

int main(int argc, char* argv[])
{
	IBNibRef nibRef;
	OSStatus err;

	// check the correct BASS was loaded
	if (HIWORD(BASS_GetVersion()) != BASSVERSION) {
		Error("An incorrect version of BASS was loaded");
		return 0;
	}

	// initalize default recording device
	if (!BASS_RecordInit(-1)) {
		Error("Can't initialize recording device");
		return 0;
	}
	// initialize default output device
	if (!BASS_Init(-1, FREQ, 0, NULL, NULL))
		Error("Can't initialize output device");

	// Create Window and stuff
	err = CreateNibReference(CFSTR("rectest"), &nibRef);
	if (err) return err;
	err = CreateWindowFromNib(nibRef, CFSTR("Window"), &win);
	if (err) return err;
	DisposeNibReference(nibRef);

	{ // get list of inputs
		int c;
		const char *i;
		ControlRef cref = GetControl(13);
		MenuRef menu = GetControlPopupMenuHandle(cref);
		DeleteMenuItem(menu, 1); // remove placeholder
		for (c = 0; i = BASS_RecordGetInputName(c); c++) {
			CFStringRef cs = CFStringCreateWithCString(0, i, 0);
			AppendMenuItemTextWithCFString(menu, cs, 0, 0, 0);
			CFRelease(cs);
			if (!(BASS_RecordGetInput(c, NULL) & BASS_INPUT_OFF)) // this 1 is currently "on"
				input = c;
		}
		if (!c) { // no input controls
			AppendMenuItemTextWithCFString(menu, CFSTR("no controls"), 0, 0, 0);
			DisableControl(cref);
		} else {
			SetControl32BitMaximum(cref, c);
			SetControl32BitValue(cref, input + 1);
			UpdateInputInfo(); // display info
		}
	}

	SetupControlHandler(10, kEventControlHit, RecordEventHandler);
	SetupControlHandler(11, kEventControlHit, PlayEventHandler);
	SetupControlHandler(12, kEventControlHit, SaveEventHandler);
	SetupControlHandler(13, kEventControlValueFieldChanged, DeviceEventHandler);
	SetControlAction(GetControl(14), NewControlActionUPP(VolumeEventHandler));

	EventLoopTimerRef timer;
	InstallEventLoopTimer(GetCurrentEventLoop(), kEventDurationNoWait, kEventDurationSecond / 5, NewEventLoopTimerUPP(TimerProc), 0, &timer);

	ShowWindow(win);
	RunApplicationEventLoop();

	// release all BASS stuff
	BASS_RecordFree();
	BASS_Free();

	return 0;
}

File: BASS_LIBS/bass24-android/examples/rectest/src/main/java/com/example/rectest/RecTest.java
/*
	BASS recording example
	Copyright (c) 2002-2019 Un4seen Developments Ltd.
*/

package com.example.rectest;

import android.Manifest;
import android.app.Activity;
import android.app.AlertDialog;
import android.os.Build;
import android.os.Bundle;
import android.os.Environment;
import android.os.Handler;
import android.view.View;
import android.widget.Button;
import android.widget.ProgressBar;
import android.widget.TextView;

import java.io.File;
import java.io.FileOutputStream;
import java.io.IOException;
import java.nio.*;
import java.nio.channels.FileChannel;

import com.un4seen.bass.BASS;

public class RecTest extends Activity {
	static final int FREQ = 44100;
	static final int CHANS = 1;
	static final int BUFSTEP = 200000;    // memory allocation unit

	int rchan; // recording channel
	int chan; // playback channel
	ByteBuffer recbuf; // recording buffer
	int level = 0;
	Handler handler = new Handler();

	class RunnableParam implements Runnable {
		Object param;

		RunnableParam(Object p) {
			param = p;
		}

		public void run() {
		}
	}

	// display error messages
	void Error(String es) {
		// get error code in current thread for display in UI thread
		String s = String.format("%s\n(error code: %d)", es, BASS.BASS_ErrorGetCode());
		runOnUiThread(new RunnableParam(s) {
			public void run() {
				new AlertDialog.Builder(RecTest.this)
						.setMessage((String) param)
						.setPositiveButton("OK", null)
						.show();
			}
		});
	}

	BASS.RECORDPROC RecordingCallback = new BASS.RECORDPROC() {
		public boolean RECORDPROC(int handle, ByteBuffer buffer, int length, Object user) {
			// buffer the data
			try {
				recbuf.put(buffer);
			} catch (BufferOverflowException e) {
				// increase buffer size
				ByteBuffer temp;
				try {
					temp = ByteBuffer.allocateDirect(recbuf.position() + length + BUFSTEP);
				} catch (Error e2) {
					runOnUiThread(new Runnable() {
						public void run() {
							Error("Out of memory!");
							StopRecording();
						}
					});
					return false;
				}
				temp.order(ByteOrder.LITTLE_ENDIAN);
				recbuf.limit(recbuf.position());
				recbuf.position(0);
				temp.put(recbuf);
				recbuf = temp;
				recbuf.put(buffer);
			}
			return true; // continue recording
		}
	};

	void StartRecording() {
		if (chan != 0) { // free old recording
			BASS.BASS_StreamFree(chan);
			chan = 0;
			findViewById(R.id.play).setEnabled(false);
			findViewById(R.id.save).setEnabled(false);
		}
		// allocate initial buffer and write the WAVE header
		recbuf = ByteBuffer.allocateDirect(BUFSTEP);
		recbuf.order(ByteOrder.LITTLE_ENDIAN);
		recbuf.put(new byte[]{'R', 'I', 'F', 'F', 0, 0, 0, 0, 'W', 'A', 'V', 'E', 'f', 'm', 't', ' ', 16, 0, 0, 0});
		recbuf.putShort((short) 1);
		recbuf.putShort((short) CHANS);
		recbuf.putInt(FREQ);
		recbuf.putInt(FREQ * CHANS * 2);
		recbuf.putShort((short) 2);
		recbuf.putShort((short) 16);
		recbuf.put(new byte[]{'d', 'a', 't', 'a', 0, 0, 0, 0});
		// start recording
		rchan = BASS.BASS_RecordStart(FREQ, CHANS, 0, RecordingCallback, 0);
		if (rchan == 0) {
			Error("Couldn't start recording");
			return;
		}
		((Button) findViewById(R.id.record)).setText("Stop");
	}

	void StopRecording() {
		BASS.BASS_ChannelStop(rchan);
		rchan = 0;
		recbuf.limit(recbuf.position());
		((Button) findViewById(R.id.record)).setText("Record");
		// complete the WAVE header
		recbuf.putInt(4, recbuf.position() - 8);
		recbuf.putInt(40, recbuf.position() - 44);
		// enable "save" button
		if (Environment.getExternalStorageState().equals(Environment.MEDIA_MOUNTED))
			findViewById(R.id.save).setEnabled(true);
		// create a stream from the recording
		chan = BASS.BASS_StreamCreateFile(recbuf, 0, recbuf.limit(), 0);
		if (chan != 0)
			findViewById(R.id.play).setEnabled(true); // enable "play" button
	}

	public void Record(View v) {
		if (rchan == 0)
			StartRecording();
		else
			StopRecording();
	}

	public void Play(View v) {
		BASS.BASS_ChannelPlay(chan, true); // play the recorded data
	}

	public void Save(View v) {
		File file = new File(Environment.getExternalStorageDirectory(), "bass.wav");
		try {
			FileChannel fc = new FileOutputStream(file).getChannel();
			recbuf.position(0);
			fc.write(recbuf);
			fc.close();
			new AlertDialog.Builder(RecTest.this)
					.setMessage("Saved to:\n" + file.toString())
					.setPositiveButton("OK", null)
					.show();
		} catch (IOException e) {
			Error("Can't save the file");
		}
	}

	@Override
	public void onCreate(Bundle savedInstanceState) {
		super.onCreate(savedInstanceState);
		setContentView(R.layout.main);

		if (Build.VERSION.SDK_INT >= 23)
			requestPermissions(new String[]{Manifest.permission.RECORD_AUDIO, Manifest.permission.WRITE_EXTERNAL_STORAGE}, 0);

		// initialize default recording device
		if (!BASS.BASS_RecordInit(-1)) {
			Error("Can't initialize recording device");
			return;
		}
		// initialize default output device
		if (!BASS.BASS_Init(-1, FREQ, 0))
			Error("Can't initialize output device");

		// timer to update the display
		Runnable timer = new Runnable() {
			public void run() {
				String text = "-";
				int lev = 0;
				if (rchan != 0) { // recording
					text = String.format("%d", BASS.BASS_ChannelGetPosition(rchan, BASS.BASS_POS_BYTE));
					lev = BASS.BASS_ChannelGetLevel(rchan);
				} else if (chan != 0) {
					if (BASS.BASS_ChannelIsActive(chan) != BASS.BASS_ACTIVE_STOPPED) { // playing
						text = String.format("%d\n%d", BASS.BASS_ChannelGetLength(chan, BASS.BASS_POS_BYTE), BASS.BASS_ChannelGetPosition(chan, BASS.BASS_POS_BYTE));
						lev = BASS.BASS_ChannelGetLevel(chan);
					} else
						text = String.format("%d", BASS.BASS_ChannelGetLength(chan, BASS.BASS_POS_BYTE));
				}
				level = level > 2000 ? level - 2000 : 0;
				if (BASS.Utils.LOWORD(lev) > level)
					level = BASS.Utils.LOWORD(lev); // check left level
				if (CHANS > 1 && BASS.Utils.HIWORD(lev) > level)
					level = BASS.Utils.HIWORD(lev); // check right level (if stereo)
				((TextView) findViewById(R.id.position)).setText(text);
				((ProgressBar) findViewById(R.id.level)).setProgress(level);
				handler.postDelayed(this, 50);
			}
		};
		handler.postDelayed(timer, 50);
	}

	@Override
	public void onDestroy() {
		// release everything
		BASS.BASS_RecordFree();
		BASS.BASS_Free();

		super.onDestroy();
	}
}